File: javascript/total_spend_test.js
import Transaction from './transaction.js';
import InvoiceItem from './invoice_item.js';
import TotalSpend from './total_spend.js';
import { assertEqual, runTests } from './test_helper.js';

function testBySpendType1() {
  const txn1 = new Transaction("Insurance premium", "Insurance", 100);
  const itm1 = new InvoiceItem("reimbursement", 100, txn1);

  const txn2 = new Transaction("Dentist appointment", "copay", 11.33);
  const itm2 = new InvoiceItem("overage", 11.33, txn2);

  const total = TotalSpend.bySpendType([itm1, itm2]);

  const expectedTotal = {"reimbursement": 100, "overage": 11.33};
  assertEqual(total, expectedTotal);
}

function testBySpendType2() {
  const txn1 = new Transaction("Insurance premium", "Insurance", 50.75);
  const itm1 = new InvoiceItem("reimbursement", 50.75, txn1);

  const txn2 = new Transaction("Dentist appointment", "copay", 91.25);
  const itm2 = new InvoiceItem("reimbursement", 49.25, txn2);
  const itm3 = new InvoiceItem("overage", 42, txn2);

  const total = TotalSpend.bySpendType([itm1, itm2, itm3]);

  const expectedTotal = {"reimbursement": 100.0, "overage": 42.0};
  assertEqual(total, expectedTotal);
}

function testBySpendTypeAndCategory() {
  const txn1 = new Transaction("Insurance premium", "Insurance", 50.75);
  const itm1 = new InvoiceItem("reimbursement", 50.75, txn1);

  const txn2 = new Transaction("Dentist appointment", "copay", 91.25);
  const itm2 = new InvoiceItem("reimbursement", 49.25, txn2);
  const itm3 = new InvoiceItem("overage", 42, txn2);

  const txn3 = new Transaction("Target purchase", "retail", 5.25);
  const itm4 = new InvoiceItem("overage", 5.25, txn3);

  const txn4 = new Transaction("Amazon purchase", "online", 10.0);
  const itm5 = new InvoiceItem("overage", 10.0, txn4);

  const total = TotalSpend.bySpendTypeAndCategory([itm1, itm2, itm3, itm4, itm5]);

  const expectedTotal = {
    [["reimbursement", "Insurance"]]: 50.75,
    [["reimbursement", "Copay"]]: 49.25,
    [["overage", "Copay"]]: 42.0,
    [["overage", "Other"]]: 15.25
  };
  assertEqual(total, expectedTotal);
}

runTests(
  testBySpendType1,
  testBySpendType2,
  testBySpendTypeAndCategory
);

File: ruby/total_spend_test.rb
require './invoice_item.rb'
require './test_helper.rb'
require './total_spend.rb'
require './transaction.rb'

include TestHelper

def test_by_spend_type1
  txn_1 = Transaction.new(name: "Insurance premium", category: "Insurance", amount: 100)
  itm_1 = InvoiceItem.new(spend_type: "reimbursement", amount: 100, transaction: txn_1)

  txn_2 = Transaction.new(name: "Dentist appointment", category: "copay", amount: 11.33)
  itm_2 = InvoiceItem.new(spend_type: "overage", amount: 11.33, transaction: txn_2)

  total = TotalSpend.by_spend_type([itm_1, itm_2])

  assert_equal(total, {"reimbursement" => 100, "overage" => 11.33})
end

def test_by_spend_type2
  txn_1 = Transaction.new(name: "Insurance premium", category: "Insurance", amount: 50.75)
  itm_1 = InvoiceItem.new(spend_type: "reimbursement", amount: 50.75, transaction: txn_1)

  txn_2 = Transaction.new(name: "Dentist appointment", category: "copay", amount: 91.25)
  itm_2 = InvoiceItem.new(spend_type: "reimbursement", amount: 49.25, transaction: txn_2)
  itm_3 = InvoiceItem.new(spend_type: "overage", amount: 42, transaction: txn_2)

  total = TotalSpend.by_spend_type([itm_1, itm_2, itm_3])

  assert_equal(total, {"reimbursement" => 100.0, "overage" => 42.0})
end

def test_by_spend_type_and_category
  txn_1 = Transaction.new(name: "Insurance premium", category: "Insurance", amount: 50.75)
  itm_1 = InvoiceItem.new(spend_type: "reimbursement", amount: 50.75, transaction: txn_1)

  txn_2 = Transaction.new(name: "Dentist appointment", category: "copay", amount: 91.25)
  itm_2 = InvoiceItem.new(spend_type: "reimbursement", amount: 49.25, transaction: txn_2)
  itm_3 = InvoiceItem.new(spend_type: "overage", amount: 42, transaction: txn_2)

  txn_3 = Transaction.new(name: "Target purchase", category: "retail", amount: 5.25)
  itm_4 = InvoiceItem.new(spend_type: "overage", amount: 5.25, transaction: txn_3)

  txn_4 = Transaction.new(name: "Amazon purchase", category: "online", amount: 10.0)
  itm_5 = InvoiceItem.new(spend_type: "overage", amount: 10.0, transaction: txn_4)

  total = TotalSpend.by_spend_type_and_category([itm_1, itm_2, itm_3, itm_4, itm_5])

  assert_equal(total, {"reimbursement" => {"Insurance" => 50.75, "Copay" => 49.25}, "overage" => {"Copay" => 42, "Other" => 15.25}})
end

run_tests(
  method(:test_by_spend_type1),
  method(:test_by_spend_type2),
  method(:test_by_spend_type_and_category),
)
